pravega/pravega
# Issue 4677: Offset/Version Table Entry Iteration

**Change log description**  
* Implement the iteration based APIs for position based `TableSegment` reads.
* This adds a new iterator which starts at a given offset/position in a `TableSegment`, and deserializes these reads into TableEntries (as apposed to the `TableBucket` approach that currently exists). 

**Purpose of the change**  
Closes #4677.

**What the code does**  

In addition to the change log description, the new `entryIterator` has the following semantics:

* If the start position provided in the `ReadTableEntriesDelta` wire command proceeds the `COMPACTION_OFFSET`, we begin starting at the `COMPACTION_OFFSET`. 
* At the time the `newIterator` is constructed, the current segment length is read, which will act as our upper bound. Meaning that even if other entries are appended, we will still consider this point as our segment end.
* The `TableEntryIterator` also uses the same `MAX_READ_SIZE` from the `PravegaRequestProcessor`.

Three new classes (related to iteration) are introduced: `TableEntryIterator`, `EntryIteratorItemImpl` and `EntryIteratorState`. When designing these classes, one of the central goals was to closely follow the general design of the previous `key/entryIterator` that already exist. I believe this may introduce some inefficiencies with you will likely want to address:

**Areas of concern**

`PravegaRequestProcessor.readTableEntriesDelta`: Conforming to the existing iterator structure, I introduce a potentially redundant packing/unpacking of our `EntryIteratorState` object for every `TableEntry` in the iteration.

`EntryIteratorItemImpl`: Unlike a `TableBucket` which may map to mutliple `TableEntry`s, each iteration will always return just one `TableEntry`. Again, using these existing structures, the `entries` field in our `EntryIteratorItem` is always a `singletonList`.

**Testing**

Besides the standard test classes for each of these new iteration classes that were introduced, you will notice two new classes: `TableContext` and `ContainerMock`. These classes were parsed out of `ContainerTableExtensionImplTests` to promote reusability.

The `TableEntryIterator` very much requires some `TableSegment` to add to and iterate over to ensure correctness.

**In Progress**

You will notice that I do not have any tests directly testing the new methods in the `PravegaRequestProcessor`. Following troubles with setting up the mocks, it is still in progress.

**How to verify it**  
(Optional: steps to verify that the changes are effective)

Run the unit tests. More involved testing is forth coming.
